From: Khalid Aziz <khalid.aziz@hp.com>
Date: Fri, 27 Apr 2012 13:00:33 -0600
Subject: PCI: disable Bus Master on PCI device shutdown
Patch-mainline: v3.5
Git-commit: b566a22c23327f18ce941ffad0ca907e50a53d41
References: bsc#920110

Disable Bus Master bit on the device in pci_device_shutdown() to ensure PCI
devices do not continue to DMA data after shutdown.  This can cause memory
corruption in case of a kexec where the current kernel shuts down and
transfers control to a new kernel while a PCI device continues to DMA to
memory that does not belong to it any more in the new kernel.

I have tested this code on two laptops, two workstations and a 16-socket
server.  kexec worked correctly on all of them.

Signed-off-by: Khalid Aziz <khalid.aziz@hp.com>
Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
(cherry picked from commit b566a22c23327f18ce941ffad0ca907e50a53d41)
Signed-off-by: Bruce Rogers <brogers@suse.com>
---
 drivers/pci/pci-driver.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/pci/pci-driver.c b/drivers/pci/pci-driver.c
index aa16193..d0b54bf 100644
--- a/drivers/pci/pci-driver.c
+++ b/drivers/pci/pci-driver.c
@@ -429,6 +429,12 @@ static void pci_device_shutdown(struct device *dev)
 		drv->shutdown(pci_dev);
 	pci_msi_shutdown(pci_dev);
 	pci_msix_shutdown(pci_dev);
+
+	/*
+	 * Turn off Bus Master bit on the device to tell it to not
+	 * continue to do DMA
+	 */
+	pci_disable_device(pci_dev);
 }
 
 #ifdef CONFIG_PM

